rustdesk-server: add the init.d file for rustdesk-server

AutoCONFIG opened the pull request:
- 添加了一个 init.d 文件，以实现在开机时自动启动 rustdesk-server。
- 由于程序本身不支持自定义工作目录，且 OpenWRT 的 init.d 不提供设置工作目录的参数，
  因此在 init 脚本中使用 cd 命令似乎是在不修改源代码前提下最佳的解决方案。
  在当前使用场景下，这种方法几乎不会影响性能。

已经部署到实际生产环境中运用，稳定运行无异常，于是重新提交pr

1715173329 commented on net/rustdesk-server/Makefile:
```diff
@@ -34,5 +34,12 @@ define Package/rustdesk-server/description
 
+define Package/rustdesk-server/install
+	$(INSTALL_DIR) $(1)/usr/bin/
+	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/bin/* $(1)/usr/bin/
```
```suggestion
	$(INSTALL_BIN) $(PKG_INSTALL_DIR)/bin/{hbbs,hbbr} $(1)/usr/bin/
```

1715173329 commented on net/rustdesk-server/files/rustdesk-server.init:
```diff
@@ -0,0 +1,26 @@
+	fi
+
+	procd_open_instance "$NAME.server"
+	procd_set_param command sh -c "cd $WORKDIR && /usr/bin/hbbs"
```
最好通过 uci 来控制 hbbs 和 hbbr 的启停还有参数（监听端口之类的），可以参考其他的软件包，比如 microsocks。

在某些情况下 hbbs 和 hbbr 可能只需要运行一个，需要能通过 uci 配置决定启动哪些服务。

AutoCONFIG replied:
感谢天灵大佬回答。因为本次提交仅仅是为了实现开机自启动，所以并没有考虑到uci配置。
将很快提交新commit来实现uci配置。

AutoCONFIG replied:
> 最好通过 uci 来控制 hbbs 和 hbbr 的启停还有参数（监听端口之类的），可以参考其他的软件包，比如 microsocks。
> 
> 在某些情况下 hbbs 和 hbbr 可能只需要运行一个，需要能通过 uci 配置决定启动哪些服务。

询问一下天灵大佬，init脚本中的防火墙配置使用uci add firewall和json_add_array firewall这两种方式哪个更好呢，我只在openwrt官方文档中找到了使用uci配置的方式，但是有不少软件包的init文件使用json_add_array firewall配置防火墙

1715173329 replied:
imo 肯定是 json_add_array firewall，应该尽量减少对自身以外的 uci 配置的修改，也尽量减少 flash 写入。

AutoCONFIG replied:
天灵大佬，我init脚本中使用了json_add_array firewall来自动配置防火墙开启对应的端口，但是并没有在防火墙中添加上对应的配置，隔壁的microsocks软件包我也试了一下，似乎也不会在防火墙中添加上对应的开启端口的配置。这部分因为相关的开发文档实在是很少，请天灵大佬指点，代码已提交请天灵大佬查看（由于是开发测试代码，虽然整体功能已经完成，但是代码没来得及重新优化逻辑，充斥着很多调试输出），测试系统为immo-24.10-rc3(release版本编译)，下面为调试输出：
![图片](https://github.com/user-attachments/assets/46d63134-d277-459e-a19f-4dda96a5636a)

1715173329 replied:
> 似乎也不会在防火墙中添加上对应的开启端口的配置

procd data 不会修改 uci 配置，fw4 重载的时候会自动读取并生成规则。

AutoCONFIG replied:
> > 似乎也不会在防火墙中添加上对应的开启端口的配置
> 
> procd data 不会修改 uci 配置，fw4 重载的时候会自动读取并生成规则。

在脚本执行结束后，我手动使用service firewall reload(restart) 和使用fw4 reload，似乎都无法使防火墙生成规则，是什么原因呢？

1715173329 replied:
procd_open_data 必须在 procd_open_instance 里面调用，不能在 open 之前，也不能在 close 之后

1715173329 commented on net/rustdesk-server/files/rustdesk-server.conf:
```diff
@@ -0,0 +1,13 @@
+config rustdesk
```
应该用 named section，例如 `config rustdesk 'global'` `config rustdesk 'hbbs'`

1715173329 commented on net/rustdesk-server/files/rustdesk-server.init:
```diff
@@ -0,0 +1,199 @@
+    echo "$hbbr_ports"
+}
+
+load_config() {
```
就这里而言，不需要单独做一个 function。

1715173329 commented on net/rustdesk-server/files/rustdesk-server.init:
```diff
@@ -0,0 +1,199 @@
+
+	config_load "$CONF"
+
+	config_get_bool _enabled "global" "enabled"
```
变量名前没必要加下划线

1715173329 commented on net/rustdesk-server/files/rustdesk-server.init:
```diff
@@ -0,0 +1,199 @@
+	printf "TCP Ports: %s\nUDP Ports: %s\n" "$TCP_PORTS" "$UDP_PORTS"
+}
+
+check_firewall_rule_exists() {
```
这里没有判断的必要，drop

1715173329 commented on net/rustdesk-server/files/rustdesk-server.init:
```diff
@@ -0,0 +1,199 @@
+#!/bin/sh /etc/rc.common
+. /usr/share/libubox/jshn.sh
```
/etc/rc.common 会自动引入此文件，不需要显式声明。

1715173329 commented on net/rustdesk-server/files/rustdesk-server.init:
```diff
@@ -0,0 +1,199 @@
+. /usr/share/libubox/jshn.sh
+
+USE_PROCD=1
+START=95
```
这类需要网络连接的一律推荐 99

1715173329 commented on net/rustdesk-server/files/rustdesk-server.init:
```diff
@@ -0,0 +1,199 @@
+
+USE_PROCD=1
+START=95
+STOP=5
```
一般可以去掉

1715173329 commented on net/rustdesk-server/files/rustdesk-server.init:
```diff
@@ -0,0 +1,199 @@
+}
+
+# Retrieve HBBS ports
+get_hbbs_ports() {
```
这些端口 func 看起来很绕，像是 ai 生成，具体要实现什么效果？

AutoCONFIG replied:
该部分是为了生成hbbr或者hbbs监听的端口序列。hbbr和hbbs都只能指定监听例如21116这种的单一端口，但是其余端口会按照这个端口为基准自动偏移（例如指定hbbs监听21118，但是实际上hbbs会监听21117 21118 21120），所以这俩函数是实现补充完整hbbr或hbbs程序监听端口序列的功能。整个程序目前是还处于调试阶段的草稿代码，只是因为防火墙的问题临时提交一下，让天灵大佬看看什么原因无法正常添加防火墙规则。后续会重新push --force

1715173329 replied:
防火墙看上面，nft list ruleset 可以看到最后生成的规则。

AutoCONFIG replied:
> 防火墙看上面，nft list ruleset 可以看到最后生成的规则。

感谢天灵大佬！

1715173329 commented on net/rustdesk-server/files/rustdesk.conf:
```diff
@@ -0,0 +1,13 @@
+config rustdesk 'global'
+	option enabled '1'
+	option auto_fw '1'
+	option workdir '/etc/rustdesk'
```
workdir 中包含必要的证书以及其它信息，应该在升级时保留。建议硬编码此目录并将其添加到 conffiles，因为也没必要随意更改这个参数。

1715173329 commented on net/rustdesk-server/Makefile:
```diff
@@ -34,5 +34,17 @@ define Package/rustdesk-server/description
 endef
 
+define Package/rustdesk-server/conffiles
+/etc/config/rustdesk
```
```suggestion
/etc/config/rustdesk
/etc/rustdesk
```

1715173329 commented on net/rustdesk-server/files/rustdesk.conf:
```diff
@@ -0,0 +1,13 @@
+config rustdesk 'global'
+	option enabled '1'
```
按目前的配置来看这个 global 大概是不必要的，各个选项应该在各自的 instance 里配置。此外按照惯例，请默认禁止服务。

1715173329 commented on net/rustdesk-server/files/rustdesk.init:
```diff
@@ -0,0 +1,97 @@
+	}
+
+	[ "$hbbr_enabled" -eq "0" ] || {
+		[ "$web_client" -eq "0" ] && HBBR_TCP_PORTS="$hbbr_port" || HBBR_TCP_PORTS="$hbbr_port $(( hbbr_port + 2 ))"
```
像这种 && ... || 最好还是改用 if else block。

1715173329 commented on net/rustdesk-server/files/rustdesk.init:
```diff
@@ -0,0 +1,97 @@
+	local HBBR_TCP_PORTS=""
+
+	config_load "$CONF"
+	config_get_bool enabled "global" "enabled"
```
bool 类型应该加上默认值，以避免出现 `sh: out of range` 的情况。

AutoCONFIG commented on net/rustdesk-server/files/rustdesk.conf:
```diff
@@ -0,0 +1,11 @@
+config firewall 'firewall'
+	option auto_fw '0'
+	option web_client '0'
```
这部分用来操作防火墙配置，虽然通常情况下搭建rustdesk服务器都会选择开启防火墙放行，但是不排除少部分用户会在内网使用或者通过vpn隧道连接内网使用，所以得到保留供用户自行配置

AutoCONFIG replied:
其余部分测试运行正常，目前测试过程中没有遇到bug或者其他问题，总体来说已经完成并按照建议修改，如果还有不足之处天灵老师敬请指出，感谢天灵大佬的耐心指导！

1715173329 replied:
我本意是想把 auto_fw 和 web_client 放到各自的 instance 里面配置，不过想想多少也有点脱裤子放屁，就维持现状吧。

1715173329 commented on net/rustdesk-server/files/rustdesk.init:
```diff
@@ -0,0 +1,104 @@
+CONF="rustdesk"
+
+start_service() {
+	local workdir="/etc/rustdesk"
```
这个放上面 WORKDIR 吧，和 CONF 一起，此外请使用 tab 缩进

AutoCONFIG replied:
已经完成修改，缩进已经全部修正。
感谢天灵老师

1715173329 commented on net/rustdesk-server/files/rustdesk.init:
```diff
@@ -0,0 +1,104 @@
+	config_get_bool auto_fw "firewall" "auto_fw" "0"
+	config_get_bool web_client "firewall" "web_client" "0"
+	config_get_bool hbbr_enabled "hbbr" "enabled" "0"
+	config_get hbbr_port "hbbr" "port"
```
目前来看并没有处理端口为空的逻辑，请给端口也加上默认值。

1715173329 commented on net/rustdesk-server/files/rustdesk.init:
```diff
@@ -0,0 +1,104 @@
+START=99
+
+CONF="rustdesk"
+workdir="/etc/rustdesk"
```
这里大写和上面保持一致

1715173329 reviewed (approved):
LGTM, thank you!

AutoCONFIG committed:
rustdesk-server: add the init.d and uci file for rustdesk-server
- Added an init.d file to enable automatic startup of rustdesk-server on boot.
- Added a uci configuration file to uniformly control the behavior of the
  rustdesk-server program and support integration with the luci application.
- Since the program itself does not support custom working directories,
  and OpenWRT's init.d does not have parameters to set the working
  directory, using `cd` in the init script seems to be the best solution
  without patching the source code. This approach will not impact
  performance under the current requirements.

Signed-off-by: Yun Wang <maoerpet@foxmail.com>

AutoCONFIG commented on net/rustdesk-server/files/rustdesk.init:
```diff
@@ -0,0 +1,104 @@
+				json_add_string src "wan"
+				json_add_string dest_port "$HBBR_TCP_PORTS"
+				json_add_string target "ACCEPT"
+				json_close_object
```
动态配置防火墙部分加上了缩进，以提升可阅读性
